jbi/queue.py

```diff
@@ -0,0 +1,315 @@
+"""This `queue` module stores Bugzilla webhook messages that we failed to sync
+to Jira.
+
+As Bugzilla sends us webhook messages, we want to eagerly accept them and
+return a `200` response so that we don't prevent it from sending new messages.
+But if we fail to sync a bug, we want to keep the message so we can retry it
+later. We also want to store any messages that might be successfuly synced, but
+were preceded by a message that wasn't synced.
+
+Classes:
+    - QueueItem: An entry in the dead letter queue, containing information
+      about the payload, timestamp, and any associated errors when attempting
+      to sync the bug.
+    - PythonException: Information about any exception that occured when
+      syncing a bug, stored along with the item.
+    - DeadLetterQueue: Class representing the dead letter queue system, providing methods
+      for adding, retrieving, and managing queue items. Supports pluggable backends.
+    - QueueBackend: Abstract base class defining the interface for a DeadLetterQueue backend.
+    - MemoryBackend: Implementation of a QueueBackend that stores messages in memory.
+    - FileBackend: Implementation of a QueueBackend that stores messages in files.
+    - InvalidQueueDSNError: Exception raised when an invalid queue DSN is provided.
+
+"""
+
+import bisect
+import itertools
+import logging
+import shutil
+import tempfile
+import traceback
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from datetime import datetime, timezone
+from functools import lru_cache
+from pathlib import Path
+from typing import Optional
+from urllib.parse import ParseResult, urlparse
+
+from dockerflow import checks
+from pydantic import BaseModel, Field
+
+from jbi import bugzilla
+from jbi.environment import DLQUrl, get_settings
+
+logger = logging.getLogger(__name__)
+
+
+class PythonException(BaseModel, frozen=True):
+    type: str
+    description: str
+    details: str
+
+    @classmethod
+    def from_exc(cls, exc: Exception):
+        return PythonException(
+            type=exc.__class__.__name__,
+            description=str(exc),
+            details="".join(traceback.format_exception(exc)),
+        )
+
+
+class QueueItem(BaseModel, frozen=True):
+    """Dead Letter Queue entry."""
+
+    timestamp: datetime = Field(default_factory=lambda: datetime.now(tz=timezone.utc))
+    payload: bugzilla.WebhookRequest
+    error: Optional[PythonException] = None
+
+    @property
+    def identifier(self):
+        return f"{self.timestamp.isoformat()}-{self.payload.bug.id}-{self.payload.event.action}-{"error" if self.error else "postponed"}"
+
+
+@lru_cache(maxsize=1)
+def get_dl_queue():
+    settings = get_settings()
+    return DeadLetterQueue(settings.dl_queue_dsn)
+
+
+class QueueBackend(ABC):
```

Comment:
Looking at this with less sinus pressurized eyes today: Is this too much abstraction? If we're always going to be running a FileBackend in deployed services, should we even allow a MemoryBackend to be used locally?  

This feels like we might be making it easier for our future selves to make mistakes and not catch them until they're deployed.

Reply:
> Is this too much abstraction?

Hmm. Yes, I think you're right. Now that I'm thinking about it, having multiple backend classes might make more sense for something like a library or if this queue was used by multiple apps with different backend types. However:

> If we're always going to be running a FileBackend in deployed services

I don't know why, but some part of me thinks we'll be moving away from our filesystem solution eventually. If / when we do, it'll be nice to rely on the API for DeadLetterQueue staying stable. 

To that end, I do think it's nice to have this interface regardless of our backend(s). It helps `DeadLetterQueue` stay stable by keeping us honest with how it expects to interact with a backend.

So I think we should have 3 classes:
- DeadLetterQueu
- the QueueBackend interface
- the FileBackend class

What do you think @leplatrem @alexcottner?

Reply:
If that makes us feel more confident about changing things in the future, I'm good with that. That's at least dropping the in-memory queue, which is something we don't have plans for using in a deployed environment. 

I try to focus on the problem in front of me and solve exactly that, but without painting myself into a corner. So as long as it's easy to pivot in the future and we aren't doing a bunch of extra work now, I'll be happy.

Reply:
I originally added the memorybackend for the tests, to be sure we start from a clean state etc.
But I'm definitely +1 on removing it for the reasons you mention. Keep the interface to maintain the shape of the composition pattern also seems a wise choice 💯